### PR TITLE
Implement callback-based Noosphere FFI

### LINKED ISSUE #510: Update Noosphere to 0.9.x non-blocking callback FFI
- Subconscious: #514 
- Noosphere: https://github.com/subconsciousnetwork/noosphere/pull/322

### PR DESCRIPTION
**Pull request by gordonbrander**
Fixes #510 

This PR implements the Noosphere 0.9.x non-blocking callback-based FFI. It wraps these APIs in async functions for ergonomic use.

### CODE REVIEW

File: xcode/Subconscious/Shared/Services/Noosphere/SphereFile.swift [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                }
+                let data = Data(bytes: contents.ptr, count: contents.len)
+                continuation.resume(returning: data)
+            }
+            return
+        }
+    }
+    
+    deinit {
+        ns_sphere_file_free(file)
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/Noosphere/SphereFile.swift:103]: Currently crashing here.

File: xcode/Subconscious/Shared/Services/Noosphere/Sphere.swift [hunk 2]
```diff
@@ -28,7 +26,7 @@ public protocol SphereProtocol {
     
     func readHeaderNames(slashlink: Slashlink) async -> [String]
     
-    func read(slashlink: Slashlink) async throws -> Memo
+    func read(slashlink: Slashlink) async throws -> MemoData
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/Noosphere/Sphere.swift:29]: Make type concrete. We were not using this customizable associatedtype anywhere.

File: xcode/Subconscious/Shared/Services/Noosphere/Sphere.swift [hunk 3]
```diff
@@ -158,13 +155,35 @@ public actor Sphere: SphereProtocol, SpherePublisherProtocol {
     public let sphere: OpaquePointer
     private let _identity: String
     
-    private init(noosphere: Noosphere, identity: String, sphere: OpaquePointer) throws {
+    /// Initialize a sphere from an already-created sphere pointer, retrieving
+    /// identity from Noosphere.
+    ///
+    /// This is a private initializer used in implementing traversal.
+    private init(
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Services/Noosphere/Sphere.swift:162]: Create private intializer that calls `ns_sphere_identity` to read identity from already-initialized sphere pointer.

Note: initializers don't support async, so initialization must always be synchronous in Swift. Implementing this private init this way is based on the assumption that as of Noosphere 0.9.x, reading sphere identity from pointer will have minimal latency (will not be blocked on writes, and will not reach out to the network).

Note however that we do only call this initializer from the async method `traverse` (https://github.com/subconsciousnetwork/subconscious/pull/514/files#diff-6deb9e69a5a4545d3d22efbc1696a7f88ab6dc55137b68618eca8063aad0f1ccR605). Calling even a blocking initializer in that context will never block the main thread (if Swift's threadpool is doing its job).

Comment (cdata) [reply xcode/Subconscious/Shared/Services/Noosphere/Sphere.swift:162]: This is reasonable. This call should never reach out to the network or block for a long period of time.

File: xcode/Subconscious/Subconscious.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -2327,8 +2335,8 @@
 			isa = XCRemoteSwiftPackageReference;
 			repositoryURL = "https://github.com/gordonbrander/ObservableStore";
 			requirement = {
-				kind = upToNextMajorVersion;
-				minimumVersion = 0.3.0;
+				kind = exactVersion;
```
Comment (gordonbrander) [inline xcode/Subconscious/Subconscious.xcodeproj/project.pbxproj:2338]: Locking in this dependency too, while I'm at it.

ObservableStore has some breaking changes in the upcoming 0.4.0. Better to upgrade when we choose to.

**Review by cdata — APPROVED** [review]
Very exciting!

**Pull request merged**